opm/grid/cpgrid/CpGridData.cpp

```diff
@@ -2509,6 +2606,65 @@ CpGridData::refinePatch(const std::array<int,3>& cells_per_dim, const std::array
             parent_to_children_cells, child_to_parent_faces, child_to_parent_cells};
 }
 
+bool CpGridData::mark(int refCount, const cpgrid::Entity<0>& element)
+{
+    if (refCount == -1) {
+        OPM_THROW(std::logic_error, "Coarsening is not supported yet.");
+    }
+    // Check the cell to be marked for refinement has no NNC (no neighbouring connections). Throw otherwise.
+    if (hasNNCs({element.index()}) && (refCount == 1)) {
+        OPM_THROW(std::logic_error, "NNC face on a cell containing LGR is not supported yet.");
+    }
+    assert((refCount == 0) || (refCount == 1)); // Do nothing (0), Refine (1), Coarsen (-1) not supported yet.
+    if (mark_.empty()) {
+        mark_.resize(this->size(0));
+    }
+    mark_[element.index()] = refCount;
+    return (mark_[element.index()] == refCount);
+}
+
+int CpGridData::getMark(const cpgrid::Entity<0>& element) const
+{
+    if (mark_.empty()) {
+        OPM_THROW(std::logic_error, "No element has been marked.");
+    }
+    else {
+        return mark_[element.index()]; // 1 refinement, 0 doing nothing, -1 coarsening (not supported yet)
+    }
+
+}
+
+bool CpGridData::preAdapt()
+{
+    // [Indirectly] Set mightVanish flags for elements that have been marked for refinement
+    if(mark_.empty()) {
+        return false;
+    }
+    else {
+        // Auxiliary bool. It will be rewritten in case an element has been marked for refinement
+        bool refinementOrCoarsening = false;
+        for (int elemIdx = 0; elemIdx <  this-> size(0); ++elemIdx) {
+            const auto& element = Dune::cpgrid::Entity<0>(*this, elemIdx, true);
+            const auto& elemMark = getMark(element);  // 1 (to be refined), 0 (do nothing), -1 (to be coarsened - not supported yet)
+            refinementOrCoarsening = refinementOrCoarsening || (elemMark != 0);
+            if (refinementOrCoarsening) {
+                break;
+            }
+        }
+        return refinementOrCoarsening;
+    }
+}
+
+bool CpGridData::adapt()
+{
+    return preAdapt();
+}
+
+void CpGridData::postAdapt()
+{
+    mark_.resize(this->size(0), 0);
```

Comment:
```suggestion
    mark_.clear();
    mark_shrink_to_fit();
```
First line is equivalent to yours but more expressive. Second one might actually free the memory (without it the underlying C-array will have the same size as before (`mark_.capacity() > mark.size()` will hold)

Reply:
I see your point. However, I might have misinterpreted the preAdapt, adapt, postAdapt. I try to illustrate y interpretation with an example. 
- Start with a grid_0 that has 10 cells. 
- Mark 1 element for refinement into 2x2x2 refined cells.  This marking process happens in grid_0 so the vector mark_ (for grid_0) has size 10.
- Call preAdapt which returns true since 1 element has been marked for refinement.
- Call adapt. This creates refined_level_1_grid with the 2x2x2 = 8 new refined cells, and the leaf_grid with the 9 coarse cells equivalent to the non-marked elements from grid_0, together with the new 8 refined cells. So the leaf_grid has 17 cells in total.
- Call postAdapt which, in my interpretation, means leaf_grid.postAdapt() where leaf_grid.mark_ is created with size 17 cells and all marks equal to 0, that might be modified in a future refine-process. 

From your comment, I get the feeling that my interpretation is wrong, and we should always mark element on the same "starting grid", in this example, grid_0. Then, mark_.clear() would clean the marks but the capacity will be 10 (total amount of cells in grid_0) and not 17 (total amount of cells on leaf_grid) This means in particular that marking cells on the leaf_grid with a next adapt() call won't work when you mark elements with indices greater than 9. 

Is my interpretation wrong? Shall we always mark element "from a fix starting grid"?

Reply:
I am sorry. I misread your code. I thought you would resize mark_ to have size zero, here. (See also my other coment above concerning mark_).

What I am wondering is whether we should have an additional array on CpGridData that always has a positive size. (mark_.size() > 0). we have had occasions where we needed to save as much memory as possible.

Maybe we could use an alterantive approach in the long run:
- Initially `mark_` has size zero
- If you call `mark()` and `mark_` has size zero, then we resize it before marking.
- In postAdapt we then do as suggested above (clear and shrink_to_fit) to reduce the memory consumption again.

I am not 100% sure what you mean by starting_grid. But you actually might have read my mind here. I am really wondering where we should store the marks. If we store them on the leaf, then we have always a very big array. An alternative is to store it only on  the level (grid). Those might have less elements than the leaf grid and we save space for recursive refinement. But let's postpone that for now.

Reply:
I see your point. So far, we do the first two steps that you describe inside mark( int, entity) definition:
 if (mark_.empty()) {
        mark_.resize(this->size(0));
    }
    mark_[element.index()] = refCount;
    return (mark_[element.index()] == refCount);

With starting_grid I mean the grid where you select and mark elements to be refine. For example, level 0 grid (GLOBAL grid). 

The code also supports (up to further debugging) that:
- You start with a level 0 grid (unique existing grid so far)
-  You mark elements for refinement
- Calling preAdapt, adapt, and postAdapt create your refined-level-1-grid and the leaf-grid-view (mixed grid with coarse and refined cells).
- If you want, now you can mark element **on the leaf-grid-view** to be refined in a second call of adapt(). You select some leaf-elements and mark them for refinement.
- Call a second time preAdapt, adapt, postAdapt. This
     - creates refined-level-2-grid 
     -  deletes "the old leaf-grid-view" where you marked (in the second marking process) elements
     - creates a new and unique leaf-grid_view taking into account level 0 (starting_grid) elements not involved in any refinement, refined cells from refined-level-1grid, and refined-level-2-grid. 

In the end, you have:
- level 0 grid (what I would call "starting_grid_1" before invoking adapt() for first time)
- level 1 refined grid (created on the first call of adapt())
- level 2 refined grid (created on the second call of adapt() )
- leaf_grid_view_2 (created on the second call of adapt(), with "starting_grid_2" the "old leaf_grid_view" created from the frist call of adapt() that now does not exist anymore, where you selected and marked elements a second time).

I hope this helps!